File: src/flock/frontend/src/types/graph.ts
// Legacy types (still used during migration for events, WebSocket handlers)
export interface Message {
  id: string;
  type: string;
  payload: any;
  timestamp: number;
  correlationId: string;
  producedBy: string;
  tags?: string[];
  visibilityKind?: string;
  partitionKey?: string | null;
  version?: number;
  isStreaming?: boolean;
  streamingText?: string;
  consumedBy?: string[];
}

// New backend API types (Phase 1 - Spec 002)
export interface GraphRequest {
  viewMode: 'agent' | 'blackboard';
  filters: GraphFilters;
  options?: GraphRequestOptions;
}

export interface GraphFilters {
  correlation_id?: string | null;
  time_range: TimeRangeFilter;
  artifactTypes: string[];
  producers: string[];
  tags: string[];
  visibility: string[];
}

export interface TimeRangeFilter {
  preset: 'last10min' | 'last5min' | 'last1hour' | 'all' | 'custom';
  start?: string | null;
  end?: string | null;
}

export interface GraphRequestOptions {
  include_statistics?: boolean;
  label_offset_strategy?: 'stack' | 'none';
  limit?: number;
}

export interface GraphSnapshot {
  generatedAt: string;
  viewMode: 'agent' | 'blackboard';
  filters: GraphFilters;
  nodes: GraphNode[];
  edges: GraphEdge[];
  statistics: GraphStatistics | null;
  totalArtifacts: number;
  truncated: boolean;
}

export interface GraphNode {
  id: string;
  type: 'agent' | 'message';
  data: Record<string, any>;
  position: { x: number; y: number };
  hidden: boolean;
}

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
  type: 'message_flow' | 'transformation';
  label?: string | null;
  data: Record<string, any>;
  markerEnd?: { type: string; width: number; height: number };
  hidden: boolean;
}

export interface GraphStatistics {
  producedByAgent: Record<string, GraphAgentMetrics>;
  consumedByAgent: Record<string, GraphAgentMetrics>;
  artifactSummary: ArtifactSummary;
}

export interface GraphAgentMetrics {
  total: number;
  byType: Record<string, number>;
}

export interface ArtifactSummary {
  total: number;
  by_type: Record<string, number>;
  by_producer: Record<string, number>;
  by_visibility: Record<string, number>;
  tag_counts: Record<string, number>;
  earliest_created_at: string;
  latest_created_at: string;
}

// Phase 1.3: Logic Operations UX - Real-time WebSocket Events
export interface CorrelationGroupUpdatedEvent {
  timestamp: string;
  agent_name: string;
  subscription_index: number;
  correlation_key: string;
  collected_types: Record<string, number>;
  required_types: Record<string, number>;
  waiting_for: string[];
  elapsed_seconds: number;
  expires_in_seconds: number | null;
  expires_in_artifacts: number | null;
  artifact_id: string;
  artifact_type: string;
  is_complete: boolean;
}

export interface BatchItemAddedEvent {
  timestamp: string;
  agent_name: string;
  subscription_index: number;
  items_collected: number;
  items_target: number | null;
  items_remaining: number | null;
  elapsed_seconds: number;
  timeout_seconds: number | null;
  timeout_remaining_seconds: number | null;
  will_flush: 'on_size' | 'on_timeout' | 'unknown';
  artifact_id: string;
  artifact_type: string;
}

// Agent logic operations state (from /api/agents endpoint + WebSocket updates)
export interface AgentLogicOperations {
  subscription_index: number;
  subscription_types: string[];
  join?: JoinSpecConfig;
  batch?: BatchSpecConfig;
  waiting_state?: LogicOperationsWaitingState;
}

export interface JoinSpecConfig {
  correlation_strategy: 'by_key';
  window_type: 'time' | 'count';
  window_value: number;
  window_unit: 'seconds' | 'artifacts';
  required_types: string[];
  type_counts: Record<string, number>;
}

export interface BatchSpecConfig {
  strategy: 'size' | 'timeout' | 'hybrid';
  size?: number;
  timeout_seconds?: number;
}

export interface LogicOperationsWaitingState {
  is_waiting: boolean;
  correlation_groups?: CorrelationGroupState[];
  batch_state?: BatchState;
}

export interface CorrelationGroupState {
  correlation_key: string;
  created_at: string;
  elapsed_seconds: number;
  expires_in_seconds: number | null;
  expires_in_artifacts: number | null;
  collected_types: Record<string, number>;
  required_types: Record<string, number>;
  waiting_for: string[];
  is_complete: boolean;
  is_expired: boolean;
}

export interface BatchState {
  created_at: string;
  elapsed_seconds: number;
  items_collected: number;
  items_target: number | null;
  items_remaining: number | null;
  timeout_seconds?: number;
  timeout_remaining_seconds?: number;
  will_flush: 'on_size' | 'on_timeout' | 'unknown';
}

File: src/flock/components/server/models/graph.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class GraphTimeRangePreset(str, Enum):
    LAST_5_MIN = "last5min"
    LAST_10_MIN = "last10min"
    LAST_1_HOUR = "last1hour"
    ALL = "all"
    CUSTOM = "custom"


class GraphTimeRange(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    preset: GraphTimeRangePreset = GraphTimeRangePreset.LAST_10_MIN
    start: datetime | None = None
    end: datetime | None = None


class GraphFilters(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    correlation_id: str | None = None
    time_range: GraphTimeRange = Field(default_factory=GraphTimeRange)
    artifact_types: list[str] = Field(default_factory=list, alias="artifactTypes")
    producers: list[str] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    visibility: list[str] = Field(default_factory=list)


class GraphRequestOptions(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    include_statistics: bool = Field(default=True, alias="includeStatistics")
    label_offset_strategy: str = Field(default="stack", alias="labelOffsetStrategy")
    limit: int = 500


class GraphRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    view_mode: Literal["agent", "blackboard"] = Field(alias="viewMode")
    filters: GraphFilters = Field(default_factory=GraphFilters)
    options: GraphRequestOptions = Field(default_factory=GraphRequestOptions)


class GraphPosition(BaseModel):
    x: float = 0.0
    y: float = 0.0


class GraphMarker(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    type: str = "arrowclosed"
    width: float = 20.0
    height: float = 20.0


class GraphNode(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    type: Literal["agent", "message"]
    data: dict[str, Any] = Field(default_factory=dict)
    position: GraphPosition | None = None
    hidden: bool = False


class GraphEdge(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    source: str
    target: str
    type: Literal[
        "message_flow", "transformation", "pending_join", "pending_batch"
    ]  # Phase 1.5: Added pending edge types
    label: str | None = None
    data: dict[str, Any] = Field(default_factory=dict)
    marker_end: GraphMarker | None = Field(default=None, alias="markerEnd")
    hidden: bool = False


class GraphAgentMetrics(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    total: int = 0
    by_type: dict[str, int] = Field(default_factory=dict, alias="byType")


class GraphStatistics(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    produced_by_agent: dict[str, GraphAgentMetrics] = Field(
        default_factory=dict, alias="producedByAgent"
    )
    consumed_by_agent: dict[str, GraphAgentMetrics] = Field(
        default_factory=dict, alias="consumedByAgent"
    )
    artifact_summary: dict[str, Any] = Field(
        default_factory=dict, alias="artifactSummary"
    )


class GraphArtifact(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    artifact_id: str = Field(alias="artifactId")
    artifact_type: str = Field(alias="artifactType")
    produced_by: str = Field(alias="producedBy")
    consumed_by: list[str] = Field(default_factory=list, alias="consumedBy")
    published_at: datetime = Field(alias="publishedAt")
    payload: dict[str, Any] = Field(default_factory=dict)
    correlation_id: str | None = Field(default=None, alias="correlationId")
    visibility_kind: str | None = Field(default=None, alias="visibilityKind")
    tags: list[str] = Field(default_factory=list)


class GraphRun(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    run_id: str = Field(alias="runId")
    agent_name: str = Field(alias="agentName")
    correlation_id: str | None = Field(default=None, alias="correlationId")
    status: Literal["active", "completed", "error"] = "active"
    consumed_artifacts: list[str] = Field(
        default_factory=list, alias="consumedArtifacts"
    )
    produced_artifacts: list[str] = Field(
        default_factory=list, alias="producedArtifacts"
    )
    duration_ms: float | None = Field(default=None, alias="durationMs")
    started_at: datetime | None = Field(default=None, alias="startedAt")
    completed_at: datetime | None = Field(default=None, alias="completedAt")
    metrics: dict[str, Any] = Field(default_factory=dict)
    error_message: str | None = Field(default=None, alias="errorMessage")


class GraphState(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    consumptions: dict[str, list[str]] = Field(default_factory=dict)
    runs: list[GraphRun] = Field(default_factory=list)
    agent_status: dict[str, str] = Field(default_factory=dict, alias="agentStatus")


class GraphSnapshot(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    generated_at: datetime = Field(alias="generatedAt")
    view_mode: Literal["agent", "blackboard"] = Field(alias="viewMode")
    filters: GraphFilters
    nodes: list[GraphNode] = Field(default_factory=list)
    edges: list[GraphEdge] = Field(default_factory=list)
    statistics: GraphStatistics | None = None
    total_artifacts: int = Field(alias="totalArtifacts", default=0)
    truncated: bool = False
